[ARM debug] test_topology_upgrade failure in `finish_writes_and_verify`: `cdc: attempted to get a stream from a generation that we know about, but weren't able to retrieve...`

[scylla-1019.log](https://github.com/scylladb/scylladb/files/14721452/scylla-1019.log)
[scylla-1020.log](https://github.com/scylladb/scylladb/files/14721453/scylla-1020.log)
[scylla-1023.log](https://github.com/scylladb/scylladb/files/14721455/scylla-1023.log)
[scylla-1027.log](https://github.com/scylladb/scylladb/files/14721456/scylla-1027.log)
[topology_experimental_raft.test_topology_upgrade.1.log](https://github.com/scylladb/scylladb/files/14721457/topology_experimental_raft.test_topology_upgrade.1.log)

```
        logging.info("Checking correctness of data in system_distributed.cdc_streams_descriptions_v2")
>       await finish_writes_and_verify()

test/topology_experimental_raft/test_topology_upgrade.py:73: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/topology/util.py:291: in finish_and_verify
    await asyncio.gather(*tasks)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    async def do_writes():
        iteration = 0
        while not stop_event.is_set():
            start_time = time.time()
            try:
>               await cql.run_async(stmt, [iteration])
E               cassandra.cluster.NoHostAvailable: ('Unable to complete the operation against any hosts', {<Host: 127.181.90.20:9042 datacenter1>: <Error from server: code=0000 [Server error] message="cdc: attempted to get a stream from a generation that we know about, but weren't able to retrieve (generation timestamp: 2024/03/22 12:29:54, write timestamp: 2024/03/22 12:29:55). Make sure that the replicas which contain this generation's data are alive and reachable from this node.">, <Host: 127.181.90.27:9042 datacenter1>: <Error from server: code=0000 [Server error] message="cdc: attempted to get a stream from a generation that we know about, but weren't able to retrieve (generation timestamp: 2024/03/22 12:29:54, write timestamp: 2024/03/22 12:29:55). Make sure that the replicas which contain this generation's data are alive and reachable from this node.">, <Host: 127.181.90.43:9042 datacenter1>: <Error from server: code=0000 [Server error] message="cdc: attempted to get a stream from a generation that we know about, but weren't able to retrieve (generation timestamp: 2024/03/22 12:29:54, write timestamp: 2024/03/22 12:29:55). Make sure that the replicas which contain this generation's data are alive and reachable from this node.">})
```

--- a/test/topology/util.py
+++ b/test/topology/util.py
@@ -12,11 +12,12 @@ import functools
 import operator
 import pytest
 import time
-from cassandra.cluster import Session  # type: ignore # pylint: disable=no-name-in-module
-from cassandra.pool import Host        # type: ignore # pylint: disable=no-name-in-module
+from cassandra.cluster import ConnectionException, ConsistencyLevel, NoHostAvailable, Session  # type: ignore # pylint: disable=no-name-in-module
+from cassandra.pool import Host                          # type: ignore # pylint: disable=no-name-in-module
+from cassandra.util import datetime_from_uuid1           # type: ignore # pylint: disable=no-name-in-module
 from test.pylib.internal_types import ServerInfo, HostID
 from test.pylib.manager_client import ManagerClient
-from test.pylib.util import wait_for, wait_for_cql_and_get_hosts, read_barrier, get_available_host
+from test.pylib.util import wait_for, wait_for_cql_and_get_hosts, read_barrier, get_available_host, unique_name
 
 
 logger = logging.getLogger(__name__)
@@ -241,6 +242,56 @@ async def check_system_topology_and_cdc_generations_v3_consistency(manager: Mana
         # Check that the contents fetched from the current host are the same as for other nodes
         assert topo_results[0] == topo_res
 
+async def start_writes_to_cdc_table(cql: Session, concurrency: int = 3):
+    logger.info(f"Starting to asynchronously write, concurrency = {concurrency}")
+
+    stop_event = asyncio.Event()
+
+    ks_name = unique_name()
+    await cql.run_async(f"CREATE KEYSPACE {ks_name} WITH replication = {{'class': 'NetworkTopologyStrategy', 'replication_factor': 3}} AND tablets = {{ 'enabled': false }}")
+    await cql.run_async(f"CREATE TABLE {ks_name}.tbl (pk int PRIMARY KEY, v int) WITH cdc = {{'enabled':true}}")
+
+    stmt = cql.prepare(f"INSERT INTO {ks_name}.tbl (pk, v) VALUES (?, 0)")
+    stmt.consistency_level = ConsistencyLevel.ONE
+
+    async def do_writes():
+        iteration = 0
+        while not stop_event.is_set():
+            start_time = time.time()
+            try:
+                await cql.run_async(stmt, [iteration])
+            except NoHostAvailable as e:
+                for _, err in e.errors.items():
+                    # ConnectionException can be raised when the node is shutting down.
+                    if not isinstance(err, ConnectionException):
+                        logger.error(f"Write started {time.time() - start_time}s ago failed: {e}")
+                        raise
+            except Exception as e:
+                logger.error(f"Write started {time.time() - start_time}s ago failed: {e}")
+                raise
+            iteration += 1
+            await asyncio.sleep(0.01)
+
+    tasks = [asyncio.create_task(do_writes()) for _ in range(concurrency)]
+
+    async def verify():
+        generations = await cql.run_async("SELECT * FROM system_distributed.cdc_streams_descriptions_v2")
+
+        stream_to_timestamp = { stream: gen.time for gen in generations for stream in gen.streams}
+
+        cdc_log = await cql.run_async(f"SELECT * FROM {ks_name}.tbl_scylla_cdc_log")
+        for log_entry in cdc_log:
+            assert log_entry.cdc_stream_id in stream_to_timestamp
+            timestamp = stream_to_timestamp[log_entry.cdc_stream_id]
+            assert timestamp <= datetime_from_uuid1(log_entry.cdc_time)
+
+    async def finish_and_verify():
+        logger.info("Stopping write workers")
+        stop_event.set()
+        await asyncio.gather(*tasks)
+        await verify()
+
+    return finish_and_verify
 
 def log_run_time(f):
     @functools.wraps(f)

--- a/test/topology/util.py
+++ b/test/topology/util.py
@@ -7,6 +7,7 @@
 Test consistency of schema changes with topology changes.
 """
 import asyncio
+import datetime
 import logging
 import functools
 import operator
@@ -191,6 +192,22 @@ async def wait_for_cdc_generations_publishing(cql: Session, hosts: list[Host], d
 
         await wait_for(all_generations_published, deadline=deadline, period=1.0)
 
+async def wait_until_last_generation_is_in_use(cql: Session):
+    topo_res = await cql.run_async("SELECT committed_cdc_generations FROM system.topology")
+    assert len(topo_res) != 0
+    generations = topo_res[0].committed_cdc_generations
+    last_generation_ts = max(gen[0] for gen in generations)
+
+    # datetime objects returned by the driver are timezone-naive and we assume they are in UTC.
+    # To subtract timestamp, we need to make sure they are both timezone-aware (or both are timezone-naive).
+    last_generation_ts = last_generation_ts.replace(tzinfo=datetime.timezone.utc)
+    now = datetime.datetime.now(datetime.timezone.utc)
+    seconds = (last_generation_ts - now).total_seconds()
+    if seconds > 0:
+        logger.info(f"Waiting {seconds} seconds for the last generation to be in use.")
+        await asyncio.sleep(seconds)
+    else:
+        logger.info(f"The last generation is already in use.")
 
 async def check_system_topology_and_cdc_generations_v3_consistency(manager: ManagerClient, hosts: list[Host]):
     assert len(hosts) != 0

--- a/test/topology_experimental_raft/test_topology_upgrade.py
+++ b/test/topology_experimental_raft/test_topology_upgrade.py
@@ -14,7 +14,8 @@ from test.pylib.manager_client import ManagerClient
 from test.pylib.util import wait_for_cql_and_get_hosts
 from test.topology.conftest import skip_mode
 from test.topology.util import log_run_time, wait_until_topology_upgrade_finishes, \
-        wait_for_cdc_generations_publishing, check_system_topology_and_cdc_generations_v3_consistency
+        wait_for_cdc_generations_publishing, check_system_topology_and_cdc_generations_v3_consistency, \
+        start_writes_to_cdc_table
 
 
 @pytest.mark.asyncio
@@ -40,6 +41,8 @@ async def test_topology_upgrade_basic(request, manager: ManagerClient):
     for host in hosts:
         status = await manager.api.raft_topology_upgrade_status(host.address)
         assert status == "not_upgraded"
+
+    finish_writes_and_verify = await start_writes_to_cdc_table(cql)
 
     logging.info("Triggering upgrade to raft topology")
     await manager.api.upgrade_to_raft_topology(hosts[0].address)
@@ -67,3 +70,6 @@ async def test_topology_upgrade_basic(request, manager: ManagerClient):
 
     logging.info("Checking consistency of data in system.topology and system.cdc_generations_v3")
     await check_system_topology_and_cdc_generations_v3_consistency(manager, hosts)
+
+    logging.info("Checking correctness of data in system_distributed.cdc_streams_descriptions_v2")
+    await finish_writes_and_verify()

--- a/test/topology_experimental_raft/test_topology_upgrade.py
+++ b/test/topology_experimental_raft/test_topology_upgrade.py
@@ -13,7 +13,7 @@ from test.pylib.rest_client import HTTPError
 from test.pylib.manager_client import ManagerClient
 from test.pylib.util import wait_for_cql_and_get_hosts
 from test.topology.conftest import skip_mode
-from test.topology.util import log_run_time, wait_until_topology_upgrade_finishes, \
+from test.topology.util import log_run_time, wait_until_last_generation_is_in_use, wait_until_topology_upgrade_finishes, \
         wait_for_cdc_generations_publishing, check_system_topology_and_cdc_generations_v3_consistency, \
         start_writes_to_cdc_table
 
@@ -21,9 +21,12 @@ from test.topology.util import log_run_time, wait_until_topology_upgrade_finishe
 @pytest.mark.asyncio
 @skip_mode('release', 'error injections are not supported in release mode')
 @log_run_time
-async def test_topology_upgrade_basic(request, manager: ManagerClient):
+async def test_topology_upgrade_basic(request, mode: str, manager: ManagerClient):
     # First, force the first node to start in legacy mode due to the error injection
-    cfg = {'error_injections_at_startup': ['force_gossip_based_join']}
+    cfg = {
+        'error_injections_at_startup': ['force_gossip_based_join'],
+        'ring_delay_ms': 15000 if mode == 'debug' else 5000,
+    }
 
     servers = [await manager.server_add(config=cfg)]
     # Disable injections for the subsequent nodes - they should fall back to
@@ -70,6 +73,11 @@ async def test_topology_upgrade_basic(request, manager: ManagerClient):
 
     logging.info("Checking consistency of data in system.topology and system.cdc_generations_v3")
     await check_system_topology_and_cdc_generations_v3_consistency(manager, hosts)
+
+    await wait_until_last_generation_is_in_use(cql)
+
+    logging.debug("Sleeping for 1 second to make sure there are writes to the CDC table in the last generation")
+    await asyncio.sleep(1)
 
     logging.info("Checking correctness of data in system_distributed.cdc_streams_descriptions_v2")
     await finish_writes_and_verify()

